domain/application/service/application.go

```diff
@@ -263,11 +273,81 @@ func (s *ApplicationService) AddUnits(ctx context.Context, name string, units ..
 }
 
 // DeleteUnit deletes the specified unit.
+// TODO(units) - rework when dual write is refactored
+// This method is called (mostly during cleanup) after a unit
+// has been removed from mongo. The mongo calls are
+// DestroyMaybeRemove, DestroyWithForce, RemoveWithForce.
 func (s *ApplicationService) DeleteUnit(ctx context.Context, unitName string) error {
 	err := s.st.DeleteUnit(ctx, unitName)
 	return errors.Annotatef(err, "deleting unit %q", unitName)
 }
 
+// EnsureUnitDead is called by the unit agent just before it terminates.
+// TODO(units): revisit his existing logic ported from mongo
+// Note: the agent only calls this method once it gets notification
+// that the unit has become dead, so there's strictly no need to call
+// this method as the unit is already dead.
+// This method is also called during cleanup from various cleanup jobs.
+// If the unit is not found, an error satisfying [applicationerrors.UnitNotFound]
+// is returned.
+func (s *ApplicationService) EnsureUnitDead(ctx context.Context, unitName string, leadershipRevoker leadership.Revoker) error {
+	err := s.st.RunAtomic(ctx, func(ctx domain.AtomicContext) error {
+		return s.ensureUnitDead(ctx, unitName, leadershipRevoker)
+	})
+	return errors.Annotatef(err, "ensuring unit %q is dead", unitName)
+}
+
+func (s *ApplicationService) ensureUnitDead(ctx domain.AtomicContext, unitName string, leadershipRevoker leadership.Revoker) (err error) {
+	defer func() {
```

Comment:
This is potentially dangerous, and there's no real need for it to be transactional.

Just revoke in the service if the unit transitioned to _dead_ successfully.

Reply:
ensureDead() is called  from RemoveUnit() and EnsureUnitDead(), so we'd need to revoke in both places.

The behaviour matches that implemented in apiserver common DeadEnsurer.

If the issue with set unit life to dead is fixed, is there still a problem?